frame/vesting/src/lib.rs

```diff
@@ -61,41 +60,139 @@ use frame_support::traits::{
+			VestingInfo { locked, per_block, starting_block }
+		}
+
+		/// Locked amount at genesis.
+		pub fn locked(&self) -> Balance {
+			self.locked
+		}
+
+		/// Amount that gets unlocked every block after `starting_block`.
+		pub fn per_block(&self) -> Balance {
+			self.per_block
+		}
+
+		/// Starting block for unlocking(vesting).
+		pub fn starting_block(&self) -> BlockNumber {
+			self.starting_block
+		}
+
+		/// Amount locked at block `n`.
+		pub fn locked_at<BlockNumberToBalance: Convert<BlockNumber, Balance>>(
+			&self,
+			n: BlockNumber,
+		) -> Balance {
+			// Number of blocks that count toward vesting
+			// Saturating to 0 when n < starting_block
+			let vested_block_count = n.saturating_sub(self.starting_block);
+			let vested_block_count = BlockNumberToBalance::convert(vested_block_count);
+			// Return amount that is still locked in vesting
+			let maybe_balance = vested_block_count.checked_mul(&self.per_block);
+			if let Some(balance) = maybe_balance {
+				self.locked.saturating_sub(balance)
+			} else {
+				Zero::zero()
+			}
+		}
+
+		/// Block number at which the schedule ends
+		pub fn ending_block<BlockNumberToBalance: Convert<BlockNumber, Balance>>(&self) -> Balance {
+			let starting_block = BlockNumberToBalance::convert(self.starting_block);
+			let duration = if self.per_block > self.locked {
+				// If `per_block` is bigger than `locked`, the schedule will end
+				// the block after starting
+				1u32.into()
+			} else if self.per_block.is_zero() {
+				// Check for div by 0 errors, which should only be from legacy
+				// vesting schedules since new ones are validated for this.
+				self.locked
+			} else {
+				let has_remainder = !(self.locked % self.per_block).is_zero();
+				let maybe_duration = self.locked / self.per_block;
+				if has_remainder {
+					maybe_duration + 1u32.into()
+				} else {
+					maybe_duration
+				}
+			};
+			starting_block.saturating_add(duration)
+		}
+	}
 }
 
-impl<
-	Balance: AtLeast32BitUnsigned + Copy,
-	BlockNumber: AtLeast32BitUnsigned + Copy,
-> VestingInfo<Balance, BlockNumber> {
-	/// Amount locked at block `n`.
-	pub fn locked_at<
-		BlockNumberToBalance: Convert<BlockNumber, Balance>
-	>(&self, n: BlockNumber) -> Balance {
-		// Number of blocks that count toward vesting
-		// Saturating to 0 when n < starting_block
-		let vested_block_count = n.saturating_sub(self.starting_block);
-		let vested_block_count = BlockNumberToBalance::convert(vested_block_count);
-		// Return amount that is still locked in vesting
-		let maybe_balance = vested_block_count.checked_mul(&self.per_block);
-		if let Some(balance) = maybe_balance {
-			self.locked.saturating_sub(balance)
-		} else {
-			Zero::zero()
+/// The indexes of vesting schedules to remove from an accounts vesting schedule collection.
+enum Filter {
+	/// Do not filter out any schedules.
+	Zero,
+	/// Filter out 1 schedule.
+	One(usize),
+	/// Filter out 2 schedules.
+	Two((usize, usize)),
```

Comment:
Why the double tuple?

```suggestion
	Two(usize, usize),
```